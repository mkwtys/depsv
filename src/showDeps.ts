import { URL } from 'url'
import chalk from 'chalk'
import Module from 'module'
import packageJson from 'package-json'
import path from 'path'
import pathExists from 'path-exists'
import readPkg from 'read-pkg'

type Dependency = {
  [pkgName: string]: {
    homepage: string
    npm: string
    repository: string
    version: string
  }
}

type Dependencies = {
  dependencies?: Dependency
  devDependencies?: Dependency
  peerDependencies?: Dependency
  optionalDependencies?: Dependency
  bundleDependencies?: Dependency
  bundledDependencies?: Dependency
}

function getPackagePath(pkgName: string, options: { cwd?: string }) {
  const cwd = options && options.cwd ? options.cwd : process.cwd()
  // @ts-ignore - Module._nodeModulePaths
  const nodeModulesPaths: string[] = Module._nodeModulePaths(cwd)
  const packagePath = path.join(nodeModulesPaths[0], pkgName)
  if (pathExists.sync(packagePath)) {
    return packagePath
  }
}

export async function createDeps({ cwd }: { cwd: string }): Promise<Dependencies> {
  const pkg = await readPkg({ cwd })

  const createSummary = async (deps): Promise<Dependency> => {
    if (!deps) {
      return {}
    }

    const homepageToString = (home: string) => {
      if (!home) {
        return ''
      }
      const h = new URL(home)
      return `${h.origin}${h.pathname}${h.search}`
    }

    const repositoryToString = (repo: string) => {
      if (!repo) {
        return ''
      }
      const r = new URL(repo)
      return `https://${r.host}${r.pathname.replace('.git', '')}`
    }

    return await Object.keys(deps).reduce(async (prev, value) => {
      const prevValue = await prev
      const pkgPath = getPackagePath(value, { cwd })
      const pkg = pkgPath ? await readPkg({ cwd: pkgPath }) : await packageJson(value)

      return {
        ...{
          [value]: {
            homepage: homepageToString(pkg.homepage),
            npm: `https://www.npmjs.com/package/${pkg.name}`,
            repository: repositoryToString(pkg.repository?.url),
            version: pkg.version,
          },
        },
        ...prevValue,
      }
    }, Promise.resolve({}))
  }

  return {
    dependencies: {
      ...(await createSummary(pkg.dependencies)),
    },
    devDependencies: {
      ...(await createSummary(pkg.devDependencies)),
    },
    peerDependencies: {
      ...(await createSummary(pkg.peerDependencies)),
    },
    optionalDependencies: {
      ...(await createSummary(pkg.optionalDependencies)),
    },
    bundleDependencies: {
      ...(await createSummary(pkg.bundleDependencies)),
    },
    bundledDependencies: {
      ...(await createSummary(pkg.bundledDependencies)),
    },
  }
}

export function depsToString(deps: Dependencies) {
  const INDENT = 2
  const PADDING = 2
  const maxDepsNameLength = Math.max.apply(
    null,
    Object.entries(deps).map(([depsName, dep]) =>
      Math.max.apply(
        null,
        Object.keys(dep).map((pkgName) => pkgName.length)
      )
    )
  )
  const maxVersionLength = Math.max.apply(
    null,
    Object.entries(deps).map(([depsName, dep]) =>
      Math.max.apply(
        null,
        Object.entries(dep).map(([pkgName, pkgSummary]) => pkgSummary.version.length)
      )
    )
  )
  const maxNpmLength = Math.max.apply(
    null,
    Object.entries(deps).map(([depsName, dep]) =>
      Math.max.apply(
        null,
        Object.entries(dep).map(([pkgName, pkgSummary]) => pkgSummary.npm.length)
      )
    )
  )
  const spaceToString = (repeat) => ' '.repeat(repeat)
  const versionIndentToString = (pkgName) => spaceToString(maxDepsNameLength - pkgName.length + PADDING)
  const urlIndentToString = (version) => spaceToString(maxVersionLength - version.length + PADDING)
  const nextUrlLineIndentToString = () =>
    spaceToString(INDENT + maxDepsNameLength + PADDING + maxVersionLength + PADDING)

  return Object.entries(deps)
    .filter(([depsName, dep]) => Object.keys(dep).length)
    .map(
      ([depsName, dep]) =>
        `\n${spaceToString(INDENT)}${chalk.green.bold(depsName)}\n${Object.entries(dep)
          .sort((a, b) => {
            if (a[0] < b[0]) {
              return -1
            }
            if (a[0] > b[0]) {
              return 1
            }
            return 0
          })
          .map(([pkgName, { homepage, repository, version, npm }]) => {
            return `${spaceToString(INDENT)}${pkgName}${versionIndentToString(pkgName)}${chalk.gray(
              version
            )}${urlIndentToString(version)}${chalk.blue.underline(npm)}${spaceToString(
              PADDING + maxNpmLength - npm.length
            )}${
              homepage && homepage !== ''
                ? chalk.blue.underline(homepage)
                : repository && repository !== '' && homepage !== repository
                ? chalk.blue.underline(repository)
                : repository === ''
                ? ''
                : `${spaceToString(INDENT)}${chalk.blue.underline(repository)}`
            }`
          })
          .join('\n')}`
    )
    .join('\n')
}

export async function showDeps(options?: { cwd?: string }) {
  const cwd = options && options.cwd ? options.cwd : process.cwd()
  const deps = await createDeps({ cwd })
  console.log(depsToString(deps))
}
