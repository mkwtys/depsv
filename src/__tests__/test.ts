import path from 'path'
import stripAnsi from 'strip-ansi'
import { createDeps, depsToString } from '../showDeps'

it('createDeps', async () => {
  const deps = await createDeps({
    cwd: path.resolve(__dirname, 'fixtures'),
  })
  expect(deps).toMatchSnapshot()
  const depsString = await depsToString(deps)
  expect(depsString).toMatchSnapshot()
  expect(stripAnsi(depsString)).toMatchSnapshot()
})
