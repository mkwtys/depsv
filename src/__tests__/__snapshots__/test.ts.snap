// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`depsToString 1`] = `
Object {
  "bundleDependencies": Object {},
  "bundledDependencies": Object {},
  "dependencies": Object {
    "chalk": Object {
      "homepage": "https://github.com/chalk/chalk",
      "npm": "https://www.npmjs.com/package/chalk",
      "repository": "https://github.com/chalk/chalk",
      "version": "4.1.1",
    },
    "core-js": Object {
      "homepage": "https://github.com/zloirock/core-js",
      "npm": "https://www.npmjs.com/package/core-js",
      "repository": "https://github.com/zloirock/core-js",
      "version": "3.15.2",
    },
    "package-json": Object {
      "homepage": "https://github.com/sindresorhus/package-json",
      "npm": "https://www.npmjs.com/package/package-json",
      "repository": "https://github.com/sindresorhus/package-json",
      "version": "7.0.0",
    },
    "path-exists": Object {
      "homepage": "https://github.com/sindresorhus/path-exists",
      "npm": "https://www.npmjs.com/package/path-exists",
      "repository": "https://github.com/sindresorhus/path-exists",
      "version": "4.0.0",
    },
    "read-pkg": Object {
      "homepage": "https://github.com/sindresorhus/read-pkg",
      "npm": "https://www.npmjs.com/package/read-pkg",
      "repository": "https://github.com/sindresorhus/read-pkg",
      "version": "6.0.0",
    },
    "yargs": Object {
      "homepage": "https://yargs.js.org/",
      "npm": "https://www.npmjs.com/package/yargs",
      "repository": "https://github.com/yargs/yargs",
      "version": "17.0.1",
    },
  },
  "devDependencies": Object {
    "@babel/cli": Object {
      "homepage": "https://babel.dev/docs/en/next/babel-cli",
      "npm": "https://www.npmjs.com/package/@babel/cli",
      "repository": "https://github.com/babel/babel",
      "version": "7.14.5",
    },
    "@babel/preset-env": Object {
      "homepage": "https://babel.dev/docs/en/next/babel-preset-env",
      "npm": "https://www.npmjs.com/package/@babel/preset-env",
      "repository": "https://github.com/babel/babel",
      "version": "7.14.7",
    },
    "@babel/preset-typescript": Object {
      "homepage": "https://babel.dev/docs/en/next/babel-preset-typescript",
      "npm": "https://www.npmjs.com/package/@babel/preset-typescript",
      "repository": "https://github.com/babel/babel",
      "version": "7.14.5",
    },
    "@types/jest": Object {
      "homepage": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jest",
      "npm": "https://www.npmjs.com/package/@types/jest",
      "repository": "https://github.com/DefinitelyTyped/DefinitelyTyped",
      "version": "26.0.24",
    },
    "@types/node": Object {
      "homepage": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node",
      "npm": "https://www.npmjs.com/package/@types/node",
      "repository": "https://github.com/DefinitelyTyped/DefinitelyTyped",
      "version": "16.3.2",
    },
    "jest": Object {
      "homepage": "https://jestjs.io/",
      "npm": "https://www.npmjs.com/package/jest",
      "repository": "https://github.com/facebook/jest",
      "version": "27.0.6",
    },
    "npm-run-all": Object {
      "homepage": "https://github.com/mysticatea/npm-run-all",
      "npm": "https://www.npmjs.com/package/npm-run-all",
      "repository": "https://github.com/mysticatea/npm-run-all",
      "version": "4.1.5",
    },
    "rimraf": Object {
      "homepage": "https://github.com/isaacs/rimraf",
      "npm": "https://www.npmjs.com/package/rimraf",
      "repository": "https://github.com/isaacs/rimraf",
      "version": "3.0.2",
    },
    "strip-ansi": Object {
      "homepage": "https://github.com/chalk/strip-ansi",
      "npm": "https://www.npmjs.com/package/strip-ansi",
      "repository": "https://github.com/chalk/strip-ansi",
      "version": "7.0.0",
    },
    "typescript": Object {
      "homepage": "https://www.typescriptlang.org/",
      "npm": "https://www.npmjs.com/package/typescript",
      "repository": "https://github.com/Microsoft/TypeScript",
      "version": "4.3.5",
    },
  },
  "optionalDependencies": Object {},
  "peerDependencies": Object {},
}
`;

exports[`depsToString 2`] = `
"
  [32m[1mdependencies[22m[39m
  chalk                     [90m4.1.1[39m    [34m[4mhttps://www.npmjs.com/package/chalk[24m[39m                     [34m[4mhttps://github.com/chalk/chalk[24m[39m
  core-js                   [90m3.15.2[39m   [34m[4mhttps://www.npmjs.com/package/core-js[24m[39m                   [34m[4mhttps://github.com/zloirock/core-js[24m[39m
  package-json              [90m7.0.0[39m    [34m[4mhttps://www.npmjs.com/package/package-json[24m[39m              [34m[4mhttps://github.com/sindresorhus/package-json[24m[39m
  path-exists               [90m4.0.0[39m    [34m[4mhttps://www.npmjs.com/package/path-exists[24m[39m               [34m[4mhttps://github.com/sindresorhus/path-exists[24m[39m
  read-pkg                  [90m6.0.0[39m    [34m[4mhttps://www.npmjs.com/package/read-pkg[24m[39m                  [34m[4mhttps://github.com/sindresorhus/read-pkg[24m[39m
  yargs                     [90m17.0.1[39m   [34m[4mhttps://www.npmjs.com/package/yargs[24m[39m                     [34m[4mhttps://yargs.js.org/[24m[39m

  [32m[1mdevDependencies[22m[39m
  @babel/cli                [90m7.14.5[39m   [34m[4mhttps://www.npmjs.com/package/@babel/cli[24m[39m                [34m[4mhttps://babel.dev/docs/en/next/babel-cli[24m[39m
  @babel/preset-env         [90m7.14.7[39m   [34m[4mhttps://www.npmjs.com/package/@babel/preset-env[24m[39m         [34m[4mhttps://babel.dev/docs/en/next/babel-preset-env[24m[39m
  @babel/preset-typescript  [90m7.14.5[39m   [34m[4mhttps://www.npmjs.com/package/@babel/preset-typescript[24m[39m  [34m[4mhttps://babel.dev/docs/en/next/babel-preset-typescript[24m[39m
  @types/jest               [90m26.0.24[39m  [34m[4mhttps://www.npmjs.com/package/@types/jest[24m[39m               [34m[4mhttps://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jest[24m[39m
  @types/node               [90m16.3.2[39m   [34m[4mhttps://www.npmjs.com/package/@types/node[24m[39m               [34m[4mhttps://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node[24m[39m
  jest                      [90m27.0.6[39m   [34m[4mhttps://www.npmjs.com/package/jest[24m[39m                      [34m[4mhttps://jestjs.io/[24m[39m
  npm-run-all               [90m4.1.5[39m    [34m[4mhttps://www.npmjs.com/package/npm-run-all[24m[39m               [34m[4mhttps://github.com/mysticatea/npm-run-all[24m[39m
  rimraf                    [90m3.0.2[39m    [34m[4mhttps://www.npmjs.com/package/rimraf[24m[39m                    [34m[4mhttps://github.com/isaacs/rimraf[24m[39m
  strip-ansi                [90m7.0.0[39m    [34m[4mhttps://www.npmjs.com/package/strip-ansi[24m[39m                [34m[4mhttps://github.com/chalk/strip-ansi[24m[39m
  typescript                [90m4.3.5[39m    [34m[4mhttps://www.npmjs.com/package/typescript[24m[39m                [34m[4mhttps://www.typescriptlang.org/[24m[39m"
`;

exports[`depsToString 3`] = `
"
  dependencies
  chalk                     4.1.1    https://www.npmjs.com/package/chalk                     https://github.com/chalk/chalk
  core-js                   3.15.2   https://www.npmjs.com/package/core-js                   https://github.com/zloirock/core-js
  package-json              7.0.0    https://www.npmjs.com/package/package-json              https://github.com/sindresorhus/package-json
  path-exists               4.0.0    https://www.npmjs.com/package/path-exists               https://github.com/sindresorhus/path-exists
  read-pkg                  6.0.0    https://www.npmjs.com/package/read-pkg                  https://github.com/sindresorhus/read-pkg
  yargs                     17.0.1   https://www.npmjs.com/package/yargs                     https://yargs.js.org/

  devDependencies
  @babel/cli                7.14.5   https://www.npmjs.com/package/@babel/cli                https://babel.dev/docs/en/next/babel-cli
  @babel/preset-env         7.14.7   https://www.npmjs.com/package/@babel/preset-env         https://babel.dev/docs/en/next/babel-preset-env
  @babel/preset-typescript  7.14.5   https://www.npmjs.com/package/@babel/preset-typescript  https://babel.dev/docs/en/next/babel-preset-typescript
  @types/jest               26.0.24  https://www.npmjs.com/package/@types/jest               https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jest
  @types/node               16.3.2   https://www.npmjs.com/package/@types/node               https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node
  jest                      27.0.6   https://www.npmjs.com/package/jest                      https://jestjs.io/
  npm-run-all               4.1.5    https://www.npmjs.com/package/npm-run-all               https://github.com/mysticatea/npm-run-all
  rimraf                    3.0.2    https://www.npmjs.com/package/rimraf                    https://github.com/isaacs/rimraf
  strip-ansi                7.0.0    https://www.npmjs.com/package/strip-ansi                https://github.com/chalk/strip-ansi
  typescript                4.3.5    https://www.npmjs.com/package/typescript                https://www.typescriptlang.org/"
`;
